[Stepper] adds stepIconBuilder property

AyushBherwani1998 opened the pull request:
## Description
Adds a new `stepIconBuilder` property in `Stepper`. The property can be used to add custom icon to step based on index or state. The property can be used for all the `StepState` except for `StepState.error`.

Fixes: #121108

Before: 
<img width="492" alt="Screenshot 2023-03-16 at 11 16 16 PM" src="https://user-images.githubusercontent.com/34301187/225707881-702a5d37-7f30-40a2-9528-466769d425ca.png">

After:
<img width="498" alt="Screenshot 2023-03-16 at 11 15 45 PM" src="https://user-images.githubusercontent.com/34301187/225707941-bebfc9ba-289f-4d0c-8110-9d906ccb7bdf.png">



## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

AyushBherwani1998 committed:
adds stepIconBuilder

AyushBherwani1998 committed:
remove local vs config file

AyushBherwani1998 committed:
Merge branch 'master' into enhancement/stepper

HansMuller committed:
Merge branch 'master' into enhancement/stepper

HansMuller reviewed (commented):
This looks like a good change. Two issues to resolve:
- Why not allow the developer to override the "icon" for the error state?
- To make it possible to override the default icons for a subset of the states:
```dart
typedef StepIconBuilder = Widget? Function(int stepIndex, StepState stepState);
```
and then only use the widget returned by `stepIconBuilder` if it's non-null; other use the default widget.

AyushBherwani1998 commented:
> Why not allow the developer to override the "icon" for the error state?

The error state have a triangle which can't fit icon inside the paint area unless the icon is way too smaller, said that if Icon is bigger it will overflow the boundary of triangle which might cause opening of new issues by users. 

> To make it possible to override the default icons for a subset of the states:

For second point, I don't think it will work unless we add `stepIconBuilder` into `Step` instead of `Stepper` class. Since, it's inside `Stepper`, it will be always non nullable if user is overriding Icon for only a single `StepState`. Let me know your thoughts @HansMuller, happy to do changes.

HansMuller commented:
This is what I was suggesting:
```dart
  Widget _buildCircleChild(int index, bool oldState) {
    final StepState state = oldState ? _oldStates[index]! : widget.steps[index].state;
    final bool isDarkActive = _isDark() && widget.steps[index].isActive;

    if (widget.stepIconBuilder != null) {
      final Widget? icon = widget.stepIconBuilder!(index, state);
      if (icon != null) {
        return icon;
      }
    }

    switch (state) {
      case StepState.indexed:
      case StepState.disabled:
        return Text(
          '${index + 1}',
          style: isDarkActive ? _kStepStyle.copyWith(color: Colors.black87) : _kStepStyle,
        );
      case StepState.editing:
        return Icon(
          Icons.edit,
          color: isDarkActive ? _kCircleActiveDark : _kCircleActiveLight,
          size: 18.0,
        );
      case StepState.complete:
        return Icon(
          Icons.check,
          color: isDarkActive ? _kCircleActiveDark : _kCircleActiveLight,
          size: 18.0,
        );
      case StepState.error:
        return const Text('!', style: _kStepStyle);
    }
  }
```

AyushBherwani1998 committed:
resolve review comments

AyushBherwani1998 committed:
Merge branch 'master' into enhancement/stepper

AyushBherwani1998 committed:
make analyzer happy

AyushBherwani1998 committed:
Merge branch 'master' into enhancement/stepper

HansMuller commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -373,6 +386,12 @@ class _StepperState extends State<Stepper> with TickerProviderStateMixin {
   Widget _buildCircleChild(int index, bool oldState) {
     final StepState state = oldState ? _oldStates[index]! : widget.steps[index].state;
     final bool isDarkActive = _isDark() && widget.steps[index].isActive;
+    if(widget.stepIconBuilder != null) {
+      final Widget? icon = widget.stepIconBuilder!(index, state);
+      if(icon != null) {
+        return icon;
+      }
+    }
```
```suggestion
final Widget? icon = widget.stepIconBuilder?.call(index, state);
if (icon != null) {
  return icon;
}
```

HansMuller commented on packages/flutter/test/material/stepper_test.dart:
```diff
@@ -1259,6 +1259,48 @@ testWidgets('Stepper custom indexed controls test', (WidgetTester tester) async
+        home: Material(
+          child: Stepper(
+            stepIconBuilder: (int index, StepState state) {
+              if(state == StepState.complete) {
```
```suggestion
              if (state == StepState.complete) {
```

HansMuller commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -303,9 +308,17 @@ class Stepper extends StatefulWidget {
+  /// Callback for creating custom icons for the [steps].
+  ///
+  /// In case of overriding icon for [StepState.error], please return
+  /// the widget having size of 14 pixels or less to avoid overflow.
```
```suggestion
  /// a widget whose width and height are 14 pixels or less to avoid overflow.
```

HansMuller commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -303,9 +308,17 @@ class Stepper extends StatefulWidget {
 
+  /// Callback for creating custom icons for the [steps].
+  ///
+  /// In case of overriding icon for [StepState.error], please return
```
```suggestion
  /// When overriding icon for [StepState.error], please return
```

HansMuller commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -106,6 +106,10 @@ class ControlsDetails {
 ///  * [WidgetBuilder], which is similar but only takes a [BuildContext].
 typedef ControlsWidgetBuilder = Widget Function(BuildContext context, ControlsDetails details);
 
+/// A builder that creates the icon widget for the step given [stepIndex], and
```
```suggestion
/// A builder that creates the icon widget a [Step] given [stepIndex], and
```

HansMuller committed:
Apply suggestions from code review

Made a few cosmetic changes; this should be ready to land once the presubmits pass.

HansMuller commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -106,6 +106,10 @@ class ControlsDetails {
 ///  * [WidgetBuilder], which is similar but only takes a [BuildContext].
 typedef ControlsWidgetBuilder = Widget Function(BuildContext context, ControlsDetails details);
 
+/// A builder that creates the icon widget a [Step] given [stepIndex], and
```
```suggestion
/// A builder that creates the icon widget for the [Step] at [stepIndex], given
```

HansMuller commented on packages/flutter/lib/src/material/stepper.dart:
```diff
@@ -373,6 +386,10 @@ class _StepperState extends State<Stepper> with TickerProviderStateMixin {
   Widget _buildCircleChild(int index, bool oldState) {
     final StepState state = oldState ? _oldStates[index]! : widget.steps[index].state;
     final bool isDarkActive = _isDark() && widget.steps[index].isActive;
+final Widget? icon = widget.stepIconBuilder?.call(index, state);
+if (icon != null) {
+  return icon;
+}
```
```suggestion
    final Widget? icon = widget.stepIconBuilder?.call(index, state);
    if (icon != null) {
      return icon;
    }
```

HansMuller committed:
Apply suggestions from code review

HansMuller committed:
Merge branch 'master' into enhancement/stepper

HansMuller committed:
Merge branch 'master' into enhancement/stepper

HansMuller reviewed (approved):
LGTM

AyushBherwani1998 commented:
Thanks for resolving the comments @HansMuller
